🦊 [AI Accelerators] Consolidate native_layer_norm for nested tensor

ani300 opened the pull request:
Summary: In order to make the layer normalization implementation for nested tensors public, it needs to be generalized to accept a normalized_shape argument instead of assuming it to be the last dimension of the nested_tensor. This commit does that, as well as adding extra unit tests to ensure the implementation is correct.

Test Plan:
All unit tests designed to test different ways of using the function work:

`buck test //caffe2/test:nested -- test_layer_norm`

Differential Revision: D40105207

drisspg commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -13131,12 +13132,6 @@
     NestedTensorCPU: NestedTensor_softmax_dropout
     NestedTensorCUDA: NestedTensor_softmax_dropout_cuda
 
-- func: _nested_tensor_layer_norm(Tensor self, Tensor? weight, Tensor? bias, float eps) -> Tensor
```
Going to have to update the backward compat I presume

drisspg commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -13131,12 +13132,6 @@
     NestedTensorCPU: NestedTensor_softmax_dropout
     NestedTensorCUDA: NestedTensor_softmax_dropout_cuda
 
-- func: _nested_tensor_layer_norm(Tensor self, Tensor? weight, Tensor? bias, float eps) -> Tensor
```
Going to have to update the backward compat probably

ani300 replied:
is this just documentation marking the function as gone or are there extra steps?

drisspg replied:
A test failure will likely pop up for from here: https://github.com/pytorch/pytorch/blob/master/test/forward_backward_compatibility/check_forward_backward_compatibility.py

drisspg commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -223,21 +275,24 @@ Tensor NestedTensor_layer_norm(
-  return at::detail::make_tensor<NestedTensorImpl>(
-      std::move(output_buffer), nt_input->get_nested_size_tensor());
+  return std::make_tuple(
+    at::detail::make_tensor<NestedTensorImpl>(std::move(output_buffer), nt_input->get_nested_size_tensor()),
```
There are other functions that are preferred to make_tensor. Specifically wrap_buffer:
https://github.com/pytorch/pytorch/blob/089a64e99e2d2b937c72e25c0fa6e4f673b8a1a1/aten/src/ATen/native/nested/NestedTensorUtils.h#L19

drisspg commented on test/test_nestedtensor.py:
```diff
@@ -381,23 +381,105 @@ def test_to_then_from_padded_tensor_no_transform0213(self, device, dtype):
+    @dtypesIfCUDA(torch.float, torch.half)
+    @skipMeta
+    @torch.inference_mode()
+    def test_layer_norm_diff_norm_shape(self, device, dtype):
```
Since this and the above tests are all shape permutations is it possible to combine into 1 iterating over the different shape combos

Commit:
🦊 [AI Accelerators] Consolidate native_layer_norm for nested tensor (#86295)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/86295

In order to make the layer normalization implementation for nested tensors public, it needs to be generalized to accept a normalized_shape argument instead of assuming it to be the last dimension of the nested_tensor. This commit does that, as well as adding extra unit tests to ensure the implementation is correct.

Test Plan:
All unit tests designed to test different ways of using the function work:

`buck test //caffe2/test:nested -- test_layer_norm`

Differential Revision: D40105207

fbshipit-source-id: fe6b9c602aaf1a115101decb384f979bf02cb224

drisspg reviewed (approved):
LGTM
